src/thor/route_action.cc

```diff
@@ -395,38 +417,33 @@ void thor_worker_t::path_arrive_by(Api& api, const std::string& costing) {
   GraphId first_edge;
   std::unordered_map<size_t, std::pair<EdgeTrimmingInfo, EdgeTrimmingInfo>> vias;
   std::vector<thor::PathInfo> path;
-  auto& correlated = *api.mutable_options()->mutable_locations();
   std::vector<std::string> algorithms;
-  api.mutable_trip()->mutable_routes()->Reserve(api.options().alternates() + 1);
+  const Options& options = api.options();
+  valhalla::Trip& trip = *api.mutable_trip();
+  trip.mutable_routes()->Reserve(options.alternates() + 1);
 
-  // For each pair of locations
-  for (auto origin = ++correlated.rbegin(); origin != correlated.rend(); ++origin) {
+  auto route_two_locations = [&, this](auto& origin, auto& destination) -> bool {
     // Get the algorithm type for this location pair
-    auto destination = std::prev(origin);
-    thor::PathAlgorithm* path_algorithm =
-        get_path_algorithm(costing, *origin, *destination, api.options());
+    thor::PathAlgorithm* path_algorithm = get_path_algorithm(costing, *origin, *destination, options);
     path_algorithm->Clear();
     algorithms.push_back(path_algorithm->name());
     LOG_INFO(std::string("algorithm::") + path_algorithm->name());
 
-    // TODO: delete this and send all cases to the function above
     // If we are continuing through a location we need to make sure we
     // only allow the edge that was used previously (avoid u-turns)
-    bool through = destination->type() == valhalla::Location::kThrough ||
-                   destination->type() == valhalla::Location::kBreakThrough;
-    while (through && first_edge.Is_Valid() && destination->path_edges_size() > 1) {
-      if (destination->path_edges().rbegin()->graph_id() == first_edge) {
-        destination->mutable_path_edges()->SwapElements(0, destination->path_edges_size() - 1);
-      }
-      destination->mutable_path_edges()->RemoveLast();
+    if (is_through_point(*destination) && first_edge.Is_Valid()) {
+      remove_path_edges(*destination,
+                        [&first_edge](const auto& edge) { return edge.graph_id() != first_edge; });
     }
 
     // Get best path and keep it
-    auto temp_paths = get_path(path_algorithm, *origin, *destination, costing, api.options());
+    auto temp_paths = get_path(path_algorithm, *origin, *destination, costing, options);
+    if (temp_paths.empty())
+      return false;
```

Comment:
nit: this line isnt covered because your unit test doesnt use date_time type=2 (arrive_by). you could increase coverage by adding another permutation just in case to make sure that the fix works for arrive_by as well.